apps/src/applab/designMode.js

```diff
@@ -721,42 +721,76 @@ function makeDraggable (jqueryElements) {
+      cancel: false,        // allow buttons and inputs to be dragged
+      appendTo: '#codeApp', // draggable can break free of design mode container
+      helper: 'clone',      // clone original ui component
+      zIndex: 2,            // put dragged item in front of design mode container
       drag: function (event, ui) {
-        // draggables are not compatible with CSS transform-scale,
-        // so adjust the position in various ways here.
-
-        // dragging
+        var container = $('#designModeViz');
         var scale = getVisualizationScale();
-        var newLeft  = ui.position.left / scale;
-        var newTop = ui.position.top / scale;
+        var newPosition = getNewPosition(container, ui.helper, scale);
+        var newLeft = newPosition.left;
+        var newTop = newPosition.top;
+        var boundsCheck = checkBoundaries(container, ui.helper, newLeft, newTop);
+        var opacity = boundsCheck.deleteElement ? 0.3 : 1;
 
-        // snap top-left corner to nearest location in the grid
-        newLeft = gridUtils.snapToGridSize(newLeft);
-        newTop = gridUtils.snapToGridSize(newTop);
-
-        // containment
-        var container = $('#designModeViz');
-        var maxLeft = container.outerWidth() - ui.helper.outerWidth(true);
-        var maxTop = container.outerHeight() - ui.helper.outerHeight(true);
-        newLeft = Math.min(newLeft, maxLeft);
-        newLeft = Math.max(newLeft, 0);
-        newTop = Math.min(newTop, maxTop);
-        newTop = Math.max(newTop, 0);
-
-        ui.position.left = newLeft;
-        ui.position.top = newTop;
-
-        elm.css({
-          top: newTop,
-          left: newLeft
-        });
+        ui.helper.css('opacity', opacity);
 
         designMode.renderDesignWorkspace(elm[0]);
       },
-      start: function () {
+      start: function (event, ui) {
         highlightElement(elm[0]);
+        ui.helper.addClass('draggingParent');
+
+        // adjust / scale cloned element
+        var scale = getVisualizationScale();
+        ui.helper.css('transform', 'scale(' + scale + ')');
+
+        // hide grip / handle when dragging
+        var grip = elm.parent();
+        grip.hide();
       },
+      stop: function(event, ui) {
+        // show original element and grip when drag stops
+        elm.css('visibility', '');
+        elm.parent().show();
+ 
+        // get outer boundaries of design mode 
+        var container = $('#designModeViz');
+
+        // get left and top of dragging element
+        var scale = getVisualizationScale();
+
+        var newPosition = getNewPosition(container, ui.helper, scale);
+        var newLeft = newPosition.left;
+        var newTop = newPosition.top;
+        var boundsCheck = checkBoundaries(container, ui.helper, newLeft, newTop);
+
+        var updatedPosition = {
+            left: newLeft - container.offset().left,
+            top: newTop - container.offset().top
+        };
```

Comment:
This code uses a mix of absolute and relative positions. I think this rename could help clarify things: `updatedPosition` --> `relativePosition`